zign/api.py

```diff
@@ -6,12 +6,87 @@
 import time
 import tokens
 import requests
+import socket
+import webbrowser
 import yaml
 
 from .config import KEYRING_KEY, TOKENS_FILE_PATH
+from oauth2client import tools
+from requests import RequestException
+from urllib.parse import parse_qs
+from urllib.parse import urlparse
+from urllib.parse import urlunsplit
 
 TOKEN_MINIMUM_VALIDITY_SECONDS = 60*5  # 5 minutes
 
+SUCCESS_PAGE = '''<!DOCTYPE HTML>
+<html lang="en-US">
+  <head>
+    <title>Authentication Successful - Zign</title>
+    <style>
+        body {
+            font-family: sans-serif;
+        }
+    </style>
+  </head>
+  <body>
+    <p>You are now authenticated with Zign.</p>
+    <p>The authentication flow has completed. You may close this window.</p>
+  </body>
+</html>'''
+
+EXTRACT_TOKEN_PAGE = '''<!DOCTYPE HTML>
+<html lang="en-US">
+  <head>
+    <!-- fill out the blanks please -->
+    <title>...</title>
+    <style>
+        body {{
+            font-family: sans-serif;
+        }}
+        #error {{
+            color: red;
+        }}
+    </style>
+    <script>
+        (function() {{
+            function parseQueryString(qs) {{
+                const result = {{}}
+                for(const part of qs.split("&")) {{
+                    const [key, val = ""] = part.split("=")
+                    result[decodeURIComponent(key)] = decodeURIComponent(val)
+                }}
+                return result
+            }}
+            const fragment = window.location.hash.substring(1)
+            const params = parseQueryString(fragment)
+            if (params.access_token) {{
+                window.location.href = "http://localhost:{port}/?" + fragment
+            }} else {{
+                document.getElementById("error").style = "display: block;"
+            }}
+        }})();
+    </script>
+  </head>
+  <body>
+    <p style="display:none;" id="error">
+        Please put a nice message here.
```

Comment:
We should also have a `<noscript>` section for people having JavaScript disabled!